### ISSUE TITLE
Observables for energy and pressure are unreliable

### ISSUE THREAD

**Issue by jngrad** [espressomd/espresso#4787]
**TL;DR:** the per-type and per-mol_id contributions from `system.analysis.energy()`, `system.analysis.pressure()` and `system.analysis.pressure_tensor()` are wrong; the total pressure is unreliable when `mol_id` properties are set.

The pressure kernels confuse molecule ids with particle types:
https://github.com/espressomd/espresso/blob/ee9d1a2dfbb7862e5d549d6a4005a0841bd46095/src/core/pressure_inline.hpp#L74-L76

The energy kernels pass particle types as argument, but molecule ids are missing:
https://github.com/espressomd/espresso/blob/ee9d1a2dfbb7862e5d549d6a4005a0841bd46095/src/core/energy_inline.hpp#L185-L188

The total energy and total pressure are correct if `mol_id` **is not used**. Individual contributions to the energy or pressure (dictionary keys with syntax `["quantity", type1, type2]` from `system.analysis`) are unreliable. Both `system.analysis` and `espressomd.observables` are affected.

This bug was introduced on June 22, 2020 in the python branch and affects releases 4.2.0 and 4.2.1. Whether the 4.0 and 4.1 releases are correct is unclear, since the energy/pressure analysis code has never been tested for `mol_id > 0`, and the per-property contributions are not correctly tested.

Here is a MWE for ESPResSo 4.2.1 (can trigger a segmentation fault and other kinds of fatal errors):

```python
import espressomd.observables
system = espressomd.System(box_l=3 * [20.], time_step=0.01)
system.cell_system.skin = 0.4
system.non_bonded_inter[0, 2].wca.set_params(epsilon=1., sigma=1.)
p0 = system.part.add(pos=[1, 2, 2], type=0, mol_id=18)
p1 = system.part.add(pos=[2, 2, 2], type=2, mol_id=18)
pressure = system.analysis.pressure()
energy = system.analysis.energy()
print(f"{pressure['total']=}")
print(f"{pressure['non_bonded']=}")
list(map(print, (f"pressure['non_bonded, {i}, {j}]={pressure['non_bonded', i, j]}" for i in range(3) for j in range(i, 3))))
print(f"{espressomd.observables.Pressure().calculate()=}")
print(f"{energy['total']=}")
print(f"{energy['non_bonded']=}")
list(map(print, (f"energy['non_bonded', {i}, {j}]={energy['non_bonded', i, j]}" for i in range(3) for j in range(i, 3))))
list(map(print, (f"energy['non_bonded_inter', {i}, {j}]={energy['non_bonded_inter', i, j]}" for i in range(3) for j in range(i, 3))))
list(map(print, (f"energy['non_bonded_intra', {i}, {j}]={energy['non_bonded_intra', i, j]}" for i in range(3) for j in range(i, 3))))
```

Annotated output:
```
pressure['total']=0.0                 <-- wrong, should be 0.001
pressure['non_bonded']=0.0            <-- wrong
pressure['non_bonded, 0, 0]=0.0
pressure['non_bonded, 0, 1]=0.0
pressure['non_bonded, 0, 2]=0.0       <-- wrong
pressure['non_bonded, 1, 1]=0.0
pressure['non_bonded, 1, 2]=0.0
pressure['non_bonded, 2, 2]=0.0
espressomd.observables.Pressure().calculate()=array([0.]) <-- wrong
energy['total']=1.0                   <-- correct
energy['non_bonded']=1.0              <-- correct
energy['non_bonded', 0, 0]=0.0
energy['non_bonded', 0, 1]=0.0
energy['non_bonded', 0, 2]=1.0        <-- correct
energy['non_bonded', 1, 1]=0.0
energy['non_bonded', 1, 2]=0.0
energy['non_bonded', 2, 2]=0.0
energy['non_bonded_inter', 0, 0]=0.0
energy['non_bonded_inter', 0, 1]=0.0
energy['non_bonded_inter', 0, 2]=1.0  <-- wrong, should be 0.0
energy['non_bonded_inter', 1, 1]=0.0
energy['non_bonded_inter', 1, 2]=0.0
energy['non_bonded_inter', 2, 2]=0.0
energy['non_bonded_intra', 0, 0]=0.0
energy['non_bonded_intra', 0, 1]=0.0
energy['non_bonded_intra', 0, 2]=0.0  <-- wrong, should be 1.0
energy['non_bonded_intra', 1, 1]=0.0
energy['non_bonded_intra', 1, 2]=0.0
energy['non_bonded_intra', 2, 2]=0.0
```

**Comment by jonaslandsgesell** [comment]
Thank you for bringing this topic up. It makes me curious: 
Given a toy system with 2 interacting particles, having type A and B (the next more complicated example would be 3 particles with types A,B,C). 
1) What is the excess pressure of particle type B (how is the contribution distributed between types A and B)? 
2) Are the excess pressure contributions additive and give the final pressure when summed up? 
3) What happens when the particles A and B do not interact with each other if type C is missing, but do interact if type C is present (e.g. due to a type based three-body interaction)?
4) What are the defining formulas for an excess pressure by type and how are they derived?

The simple example with two particles illustrates that only the interplay of two (or more) particles (and their types) have an excess pressure but not the particle type alone. 

Side note: it is possible to do such kind of assignments of partial contributions, e.g. in game theory: https://en.m.wikipedia.org/wiki/Shapley_value a game would be a simulated system and a player would be a particle type, a value of a coalition is the observed pressure of the system.

**Comment by jngrad** [comment]
@jonaslandsgesell We use the virial pressure. Therefore we don't calculate the excess pressure per particle types, but per particle pairs or triplets. More details are available in the user guide section on [pressure](https://espressomd.github.io/doc4.2.1/analysis.html#pressure).
